File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int maxLength=0;
        int n=s.length();
        unordered_map<char,int> charSet;
        int left=0;
    
        for(int right=0;right<n;right++)
        {
            if(charSet.count(s[right])==0 || charSet[s[right]]<left)
            {
                charSet[s[right]]=right;
                maxLength=max(maxLength,right-left+1);
            }
            else 
            {
                left=charSet[s[right]]+1;
                charSet[s[right]]=right;    
            }
        }
        return maxLength;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character,Integer> map = new HashMap<>();
        int max = 0;
        int left = 0;

        for(int right=0;right<s.length();right++){
            if(!map.containsKey(s.charAt(right)) || map.get(s.charAt(right)) < left){
                map.put(s.charAt(right),right);
                max = Math.max(max,right-left+1);
            }
            else
            {
                left = map.get(s.charAt(right)) + 1;
                map.put(s.charAt(right),right);
            }
        }
        return max;
        
    }
}